### PR TITLE
Add Dense reduction kernels using temporary Array

### PR DESCRIPTION
**Pull request by upsj**
This adds another version of run_kernel_reduction that reuses an existing array for temporary storage to avoid reallocations. The same parameter is being exposed via Dense reduction member functions. The PR also unifies the run_kernel_* implementations between CUDA and HIP.

~~I removed the sparselib kernels for simplicity, I am currently testing whether we can improve the tuning parameters of the run_kernel_reduction kernels to get rid of them entirely.~~

### CODE REVIEW

**Comment by pratikvn** [comment]
Is the plan to remove the sparselib kernel here and merge the tuning PR first ?  Maybe we keep it until the tuning has been completed, because it looks like that PR is in an incomplete state.

**Comment by upsj** [comment]
Yes, I was kind of lazy there and hoped I would manage to easily fit the tuning into this PR. I'll revert the changes

File: hip/test/base/kernel_launch.hip.cpp [hunk 1]
```diff
@@ -350,67 +412,99 @@ TEST_F(KernelLaunch, Reduction2D) { run2d_reduction(exec); }
...
-            return (i + 1) * (j + 1);
-        },
-        [] GKO_KERNEL(auto i, auto j) { return i + j; },
-        [] GKO_KERNEL(auto j) { return j * 2; }, int64{}, output.get_data(),
-        gko::dim<2>{static_cast<size_type>(num_rows),
-                    static_cast<size_type>(num_cols)},
-        output);
+    // empty, most threads idle, most threads busy, multiple blocks
+    for (auto num_rows : {0, 10, 100, 1000, 10000}) {
+        // check different edge cases: subwarp sizes, blocked mode
```
Comment (pratikvn) [inline hip/test/base/kernel_launch.hip.cpp:466]: Nice!

File: test/matrix/dense_kernels.cpp [hunk 2]
```diff
@@ -998,18 +998,135 @@ TEST_F(Dense, ExtractDiagonalOnShortFatIntoDenseCrossExecutor)
...
+
+
+TEST_F(Dense, ComputeDotWithTmpIsEquivalentToRef)
+{
+    set_up_vector_data(40);
+
+    auto dot_size = gko::dim<2>{1, x->get_size()[1]};
+    auto dot_expected = Mtx::create(ref, dot_size);
+    auto ddot = Mtx::create(ref, dot_size);
+    gko::Array<char> tmp{ref};
```
Comment (pratikvn) [inline test/matrix/dense_kernels.cpp:1024]: Maybe have a test with properly allocated `tmp` ?

**Review by pratikvn — COMMENTED** [review]
Nice work! Looks mostly good. Some minor comments.

File: include/ginkgo/core/matrix/dense.hpp [hunk 3]
```diff
@@ -1012,64 +1012,64 @@ class Dense
     /**
      * @copydoc scale(const LinOp *)
      *
-     * @note  Other implementations of dense should override this function
-     *        instead of scale(const LinOp *alpha).
+     * @deprecated  This function will be removed in the future,
```
[the remark this answers is not in the document]
Comment (upsj) [reply include/ginkgo/core/matrix/dense.hpp:1015]: Yeah, I was also a bit unsure about this. My argument here is: Even if you derive from dense, every class taking Dense inputs will use `values_` and `stride_` to determine how the data will be accessed, so there is nothing sensible behavior-wise that a user could add here. It might make sense to add a logger event to member functions like these (not individual events but a generic linop_member_called (the name is bad, I know 😆 ) with a string parameter for the actual function name.

File: core/matrix/dense.cpp [hunk 4]
```diff
@@ -269,6 +334,22 @@ void Dense<ValueType>::sub_scaled_impl(const LinOp* alpha, const LinOp* b)
...
+    auto local_res = make_temporary_clone(exec, result);
+    auto dense_b = make_temporary_conversion<ValueType>(local_b.get());
+    auto dense_res = make_temporary_conversion<ValueType>(local_res.get());
+    exec->run(
+        dense::make_compute_dot(this, dense_b.get(), dense_res.get(), tmp));
+}
+
+
 template <typename ValueType>
 void Dense<ValueType>::compute_dot_impl(const LinOp* b, LinOp* result) const
```
[the remark this answers is not in the document]
Comment (upsj) [reply core/matrix/dense.cpp:354]: yes, it's a virtual function, so we can't simply remove it for interface reasons

File: common/cuda_hip/base/kernel_launch_reduction.hpp.inc [hunk 5]
```diff
@@ -0,0 +1,534 @@
...
+                                 ValueType* result, size_type size,
+                                 Array<char>& tmp, KernelArgs&&... args)
+{
+    constexpr int oversubscription = 16;
+    constexpr auto block_size = default_block_size;
+    const auto num_blocks = std::min<int64>(
+        ceildiv(size, block_size), exec->get_num_warps() * oversubscription);
+    if (num_blocks > 1) {
+        const auto required_storage = sizeof(ValueType) * num_blocks;
+        tmp.set_executor(exec);
```
[the remark this answers is not in the document]
Comment (upsj) [reply common/cuda_hip/base/kernel_launch_reduction.hpp.inc:133]: I could do this either on the `core` side or the kernel side. If users pass in an Array on the wrong executor, I need to somehow turn it into a correct one. Either by checking on the host side and creating a different array if necessary, or by changing the executor of the input array. Also by set_executor we might copy over data, so `tmp.clear(); tmp.set_executor(exec);` might be more appropriate?

[the remark this answers is not in the document]
Comment (pratikvn) [reply common/cuda_hip/base/kernel_launch_reduction.hpp.inc:133]: I mean this kind of transparently transfers the data to the executor, with allocations, which the user probably did not intend to do, if they actually used the advanced version of the `_impl` function where they explicitly passed in a temp array. So, I am suggesting we throw instead of continue if there is an executor mismatch.

[the remark this answers is not in the document]
Comment (upsj) [reply common/cuda_hip/base/kernel_launch_reduction.hpp.inc:133]: I think this is probably not the right point to change the executor, so I'll move the operation to `core`. If `tmp` was initialized without executor, we would still need to add some special-case treatment code. How about `if (exec != array.get_executor()) { array.clear(); array.set_executor(exec); }`?

File: common/cuda_hip/base/kernel_launch_reduction.hpp.inc [hunk 6]
```diff
@@ -0,0 +1,534 @@
...
+        if (tmp.get_num_elems() < required_storage) {
+            tmp.resize_and_reset(required_storage);
```
[the remark this answers is not in the document]
Comment (pratikvn) [reply common/cuda_hip/base/kernel_launch_reduction.hpp.inc:135]: Yeah, I guess that is true. That is probably enough.

File: core/matrix/dense.cpp [hunk 7]
```diff
@@ -269,6 +334,26 @@ void Dense<ValueType>::sub_scaled_impl(const LinOp* alpha, const LinOp* b)
...
+    if (tmp.get_executor() != exec) {
+        tmp.clear();
+        tmp.set_executor(exec);
+    }
+    auto local_b = make_temporary_clone(exec, b);
+    auto local_res = make_temporary_clone(exec, result);
+    auto dense_b = make_temporary_conversion<ValueType>(local_b.get());
+    auto dense_res = make_temporary_conversion<ValueType>(local_res.get());
+    exec->run(
+        dense::make_compute_dot(this, dense_b.get(), dense_res.get(), tmp));
```
Comment (pratikvn) [inline core/matrix/dense.cpp:353]: `make_compute_dot_dispatch` ? Same for other functions.

File: cuda/test/base/kernel_launch.cu [hunk 8]
```diff
@@ -51,15 +51,42 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include "core/test/utils.hpp"
 
 
-namespace {
-
-
```
Comment (yhmtsai) [inline cuda/test/base/kernel_launch.cu:56]: the namespace is deleted

File: dpcpp/solver/cb_gmres_kernels.dp.cpp [hunk 9]
```diff
@@ -1006,13 +1006,14 @@ void initialize_2(std::shared_ptr<const DpcppExecutor> exec,
...
     const auto stride_arnoldi = arnoldi_norm->get_stride();
+    Array<char> tmp{exec};
 
     initialize_2_1_kernel<block_size>(
         grid_dim_1, block_dim, 0, exec->get_queue(), residual->get_size()[0],
         residual->get_size()[1], krylov_dim, krylov_bases,
         residual_norm_collection->get_values(),
         residual_norm_collection->get_stride());
-    kernels::dpcpp::dense::compute_norm2(exec, residual, residual_norm);
+    kernels::dpcpp::dense::compute_norm2(exec, residual, residual_norm, tmp);
```
Comment (yhmtsai) [inline dpcpp/solver/cb_gmres_kernels.dp.cpp:1016]: same here

File: dpcpp/test/base/kernel_launch.dp.cpp [hunk 10]
```diff
@@ -51,15 +51,42 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include "core/test/utils.hpp"
 
 
-namespace {
-
-
```
Comment (yhmtsai) [inline dpcpp/test/base/kernel_launch.dp.cpp:56]: namespace

File: hip/test/base/kernel_launch.hip.cpp [hunk 11]
```diff
@@ -51,15 +51,42 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include "core/test/utils.hpp"
 
 
-namespace {
-
-
```
Comment (yhmtsai) [inline hip/test/base/kernel_launch.hip.cpp:56]: namespace

File: include/ginkgo/core/matrix/dense.hpp [hunk 12]
```diff
@@ -786,12 +784,19 @@ class Dense
      *                (the number of columns in the vector must match the number
      *                of columns of this)
      */
-    void compute_norm2(LinOp* result) const
-    {
-        auto exec = this->get_executor();
-        this->compute_norm2_impl(
-            make_temporary_output_clone(exec, result).get());
-    }
+    void compute_norm2(LinOp* result) const;
```
Comment (yhmtsai) [inline include/ginkgo/core/matrix/dense.hpp:787]: this is not virtual, so you can make the tmp = {} with default?

File: omp/base/kernel_launch_reduction.hpp [hunk 13]
```diff
@@ -197,7 +204,8 @@ void run_kernel_row_reduction_impl(std::shared_ptr<const OmpExecutor> exec,
                                    KernelFunction fn, ReductionOp op,
                                    FinalizeOp finalize, ValueType identity,
                                    ValueType* result, size_type result_stride,
-                                   dim<2> size, MappedKernelArgs... args)
+                                   dim<2> size, Array<char>& tmp,
+                                   MappedKernelArgs... args)
```
Comment (yhmtsai) [inline omp/base/kernel_launch_reduction.hpp:208]: also &&?

File: omp/test/base/kernel_launch.cpp [hunk 14]
```diff
@@ -51,15 +51,42 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include "core/test/utils.hpp"
 
 
-namespace {
-
-
```
Comment (yhmtsai) [inline omp/test/base/kernel_launch.cpp:56]: namespace

File: test/matrix/dense_kernels.cpp [hunk 15]
```diff
@@ -998,18 +998,203 @@ TEST_F(Dense, ExtractDiagonalOnShortFatIntoDenseCrossExecutor)
...
+    dx->compute_conj_dot(y.get(), ddot.get(), tmp);
+
+    GKO_ASSERT_MTX_NEAR(ddot, dot_expected, r<vtype>::value);
+}
+
+
 TEST_F(Dense, ComputeNorm1IsEquivalentToRef)
 {
-    set_up_apply_data();
+    set_up_vector_data(2);
```
Comment (yhmtsai) [inline test/matrix/dense_kernels.cpp:1103]: it should keep () to test one vector

File: test/matrix/dense_kernels.cpp [hunk 16]
```diff
@@ -998,18 +998,203 @@ TEST_F(Dense, ExtractDiagonalOnShortFatIntoDenseCrossExecutor)
 }
 
 
+TEST_F(Dense, ComputeDotIsEquivalentToRef)
+{
+    set_up_vector_data(2);
```
Comment (yhmtsai) [inline test/matrix/dense_kernels.cpp:1003]: need () to test one vector

File: test/matrix/dense_kernels.cpp [hunk 17]
```diff
@@ -998,18 +998,203 @@ TEST_F(Dense, ExtractDiagonalOnShortFatIntoDenseCrossExecutor)
...
+    x->compute_norm1(norm_expected.get());
+    dx->compute_norm1(dnorm.get(), tmp);
+
+    GKO_ASSERT_MTX_NEAR(norm_expected, dnorm, r<vtype>::value);
+}
+
+
+TEST_F(Dense, ComputeNorm2IsEquivalentToRef)
+{
+    set_up_vector_data(5);
```
Comment (yhmtsai) [inline test/matrix/dense_kernels.cpp:1153]: same here

File: omp/base/kernel_launch_reduction.hpp [hunk 18]
```diff
@@ -290,7 +298,7 @@ void run_kernel_col_reduction_sized_impl(
     syn::value_list<int, remainder_cols>,
     std::shared_ptr<const OmpExecutor> exec, KernelFunction fn, ReductionOp op,
     FinalizeOp finalize, ValueType identity, ValueType* result, dim<2> size,
-    MappedKernelArgs... args)
+    Array<char>& tmp, MappedKernelArgs... args)
```
Comment (yhmtsai) [inline omp/base/kernel_launch_reduction.hpp:301]:
```suggestion
    Array<char>& tmp, MappedKernelArgs&&... args)
```

Comment (upsj) [reply omp/base/kernel_launch_reduction.hpp:301]: this is deliberate, as MappedKernelArgs need to be value types, not references

File: cuda/test/base/kernel_launch.cu (hunk 8 above)
Comment (upsj) [reply cuda/test/base/kernel_launch.cu:56]: this is an executable, not a library, so the unnamed namespace has no effect on the execution of the program

File: include/ginkgo/core/matrix/dense.hpp (hunk 12 above)
Comment (upsj) [reply include/ginkgo/core/matrix/dense.hpp:787]: Keeping them separate makes it easier to write separate documentation for them. Also I'm not sure you can have a non-const lvalue reference parameter with default value anyways.

File: test/matrix/dense_kernels.cpp (hunk 16 above)
Comment (upsj) [reply test/matrix/dense_kernels.cpp:1003]: good point, at least for (conj_)dot and norm2 we should test 1 rhs with stride == 1

File: test/matrix/dense_kernels.cpp (hunk 15 above)
Comment (upsj) [reply test/matrix/dense_kernels.cpp:1103]: this uses common kernels, so its behavior for 1 rhs is already being tested elsewhere.

**Comment by upsj** [comment]
@yhmtsai Different executors use different amounts of temporary storage, so I don't think it makes sense to reuse the previous size. It is basically intended as a way for the user to provide an opaque array that they themselves will not touch.

File: cuda/test/base/kernel_launch.cu (hunk 8 above)
Comment (yhmtsai) [reply cuda/test/base/kernel_launch.cu:56]: I know. It is from gtest example and we have it everywhere in ginkgo. but it is not important.

**Comment by yhmtsai** [comment]
@upsj I think cuda/hip and maybe dpcpp have the same amount of workspace?
omp and reference might not need it.

File: core/matrix/dense.cpp [hunk 19]
```diff
@@ -291,8 +397,25 @@ void Dense<ValueType>::compute_conj_dot_impl(const LinOp* b,
     auto exec = this->get_executor();
     auto dense_b = make_temporary_conversion<ValueType>(b);
     auto dense_res = make_temporary_conversion<ValueType>(result);
-    exec->run(dense::make_compute_conj_dot_dispatch(this, dense_b.get(),
-                                                    dense_res.get()));
+    Array<char> tmp{exec};
+    exec->run(dense::make_compute_conj_dot(this, dense_b.get(), dense_res.get(),
```
Comment (yhmtsai) [inline core/matrix/dense.cpp:401]: dispatch?

File: core/matrix/dense.cpp (hunk 7 above)
Comment (yhmtsai) [reply core/matrix/dense.cpp:353]: is it missing dispatch?

**Review by yhmtsai — APPROVED** [review]
some dispatch are still missing?

File: core/matrix/dense.cpp (hunk 7 above)
Comment (upsj) [reply core/matrix/dense.cpp:353]: no, the corresponding operation calls _dispatch, see the beginning of the file. Do you think I should change it here as well? I wanted to keep the diff small, but that of course may create some confusion. I'd assume once #991 is executed and merged, we could probably reverse this?

**Comment by upsj** [comment]
@yhmtsai The amount of storage depends on device characteristics, so it is most likely different even between different CudaExecutors, let alone between CUDA and HIP/SYCL/OpenMP. Only Reference uses no temporary storage at all.

File: cuda/test/base/kernel_launch.cu (hunk 8 above)
Comment (upsj) [reply cuda/test/base/kernel_launch.cu:56]: If there is no reason for adding it, I don't think we need to keep it. [This SO question](https://stackoverflow.com/questions/47861534/why-does-google-test-sample-put-tests-in-an-anonymous-namespace) discusses the issue, but I don't really think the answer applies to us at all. Symbol visibility is a problem that we will deal with in the future #715, but it does not have any impact here (except for maybe a miniscule difference in link time because the symbol table of the object files is bigger)

File: core/matrix/dense.cpp (hunk 7 above)
Comment (yhmtsai) [reply core/matrix/dense.cpp:353]: I see. I do not notice the function is point to the dispatch one.
it should be okay
